controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -1118,10 +1118,13 @@ private boolean isStreamCutValidInternal(Map<Long, Long> streamCut, int epochLow
                         ImmutableList<StreamSegmentRecord> newSegments = newSegmentsBuilder.build();
                         builder.addAll(newSegments);
                         // epoch record
-                        EpochRecord epochRecord = new EpochRecord(epochTransition.getNewEpoch(), epochTransition.getNewEpoch(), 
-                                builder.build(), time);
-
-                        HistoryTimeSeriesRecord timeSeriesRecord = 
+                        SimpleEntry<Long, Long> splitMergeCountLastEpoch = getSplitMergeCountsTillEpoch(currentEpoch).join();
```

Comment:
we should not be blocking here.

Reply:
we should get the future and add callbacks to perform any subsequent processing.. 
we use join in the code only in following pattern:
create future1 and future2..
```
future1 = getFuture();
future2 = getFuture();
CompletableFuture.allOf(future1, future2) 
    .thenApply(v -> {// chain subsequent processing to completion of "both" futures.
           result1 = future1.join(); // this join is going to return the result immediately without blocking as the result is available with future1 which is completed for sure.
           result2 = future2.join(); // same here
           // use my results to do some computation.
     });
```

Reply:
fixed as suggested.